### ISSUE TITLE
Red Hat 6 and CentOS 6 are no longer fully supported after version 2017.x.x

### ISSUE THREAD

**Issue by jryberg** [saltstack/salt#42979]
### Description of Issue/Question
I have been using https://github.com/saltstack-formulas/mysql-formula for ages but after I upgraded to 2017.7.0 and later 2017.7.1 it's no longer possible to manage users in MySQL for Red Hat 6(.9)

I changed the state to it's minimum and it still fails exactly the same way (took example just to be sure I did not do anything wrong: https://docs.saltstack.com/en/latest/ref/states/all/salt.states.mysql_user.html)
State test.sls:
```
frank:
  mysql_user.present:
    - host: localhost
    - password: bobcat
    - connection_host: '127.0.0.1'
    - connection_user: 'root'
    - connection_pass: '[[ REDACTED ]]'
```

Debug log

> [DEBUG   ] Reading configuration from /etc/salt/minion
> [DEBUG   ] Including configuration from '/etc/salt/minion.d/_schedule.conf'
> [DEBUG   ] Reading configuration from /etc/salt/minion.d/_schedule.conf
> [DEBUG   ] Configuration file path: /etc/salt/minion
> [WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
> [DEBUG   ] Reading configuration from /etc/salt/minion
> [DEBUG   ] Including configuration from '/etc/salt/minion.d/_schedule.conf'
> [DEBUG   ] Reading configuration from /etc/salt/minion.d/_schedule.conf
> [DEBUG   ] Connecting to master. Attempt 1 of 1
> [DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', 'aspod-johan-main01', 'tcp://salt:4506')
> [DEBUG   ] Generated random reconnect delay between '1000ms' and '11000ms' (7438)
> [DEBUG   ] Setting zmq_reconnect_ivl to '7438ms'
> [DEBUG   ] Setting zmq_reconnect_ivl_max to '11000ms'
> [DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'aspod-johan-main01', 'tcp://salt:4506', 'clear')
> [DEBUG   ] Decrypting the current master AES key
> [DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
> [DEBUG   ] SaltEvent PUB socket URI: /var/run/salt/minion/minion_event_b760395531_pub.ipc
> [DEBUG   ] SaltEvent PULL socket URI: /var/run/salt/minion/minion_event_b760395531_pull.ipc
> [DEBUG   ] Initializing new IPCClient for path: /var/run/salt/minion/minion_event_b760395531_pull.ipc
> [DEBUG   ] Sending event: tag = salt/auth/creds; data = {'_stamp': '2017-08-16T14:20:17.232240', 'creds': {'publish_port': 4505, 'aes': '[[ REDACTED ]]', 'master_uri': 'tcp://salt:4506'}, 'key': ('/etc/salt/pki/minion', 'aspod-johan-main01', 'tcp://salt:4506')}
> [DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
> [DEBUG   ] Determining pillar cache
> [DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'aspod-johan-main01', 'tcp://salt:4506', 'aes')
> [DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', 'aspod-johan-main01', 'tcp://salt:4506')
> [DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
> [DEBUG   ] LazyLoaded jinja.render
> [DEBUG   ] LazyLoaded yaml.render
> [DEBUG   ] LazyLoaded state.sls
> [DEBUG   ] LazyLoaded saltutil.is_running
> [DEBUG   ] LazyLoaded grains.get
> [DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'aspod-johan-main01', 'tcp://salt:4506', 'aes')
> [DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', 'aspod-johan-main01', 'tcp://salt:4506')
> [INFO    ] Loading fresh modules for state activity
> [DEBUG   ] LazyLoaded jinja.render
> [DEBUG   ] LazyLoaded yaml.render
> [DEBUG   ] In saltenv 'base', looking at rel_path 'test.sls' to resolve 'salt://test.sls'
> [DEBUG   ] In saltenv 'base', ** considering ** path '/var/cache/salt/minion/files/base/test.sls' to resolve 'salt://test.sls'
> [DEBUG   ] Fetching file from saltenv 'base', ** attempting ** 'salt://test.sls'
> [DEBUG   ] No dest file found
> [INFO    ] Fetching file from saltenv 'base', ** done ** 'test.sls'
> [DEBUG   ] compile template: /var/cache/salt/minion/files/base/test.sls
> [DEBUG   ] Jinja search path: ['/var/cache/salt/minion/files/base']
> [PROFILE ] Time (in seconds) to render '/var/cache/salt/minion/files/base/test.sls' using 'jinja' renderer: 0.00320887565613
> [DEBUG   ] Rendered data from file: /var/cache/salt/minion/files/base/test.sls:
> frank:
>   mysql_user.present:
>     - host: localhost
>     - password: bobcat
>     - connection_host: '127.0.0.1'
>     - connection_user: 'root'
>     - connection_pass: '[[ REDACTED ]]'
> 
> [DEBUG   ] LazyLoaded config.get
> [DEBUG   ] Results of YAML rendering:
> OrderedDict([('frank', OrderedDict([('mysql_user.present', [OrderedDict([('host', 'localhost')]), OrderedDict([('password', 'bobcat')]), OrderedDict([('connection_host', '127.0.0.1')]), OrderedDict([('connection_user', 'root')]), OrderedDict([('connection_pass', '[[ REDACTED ]]')])])]))])
> [PROFILE ] Time (in seconds) to render '/var/cache/salt/minion/files/base/test.sls' using 'yaml' renderer: 0.00501585006714
> [DEBUG   ] Failed to import module pagerduty_util:
> Traceback (most recent call last):
>   File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1411, in _load_module
>     mod = imp.load_module(mod_namespace, fn_, fpath, desc)
>   File "/usr/lib/python2.7/site-packages/salt/modules/pagerduty_util.py", line 24, in <module>
>     import requests
>   File "/usr/lib/python2.7/site-packages/requests/__init__.py", line 58, in <module>
>     from . import utils
>   File "/usr/lib/python2.7/site-packages/requests/utils.py", line 25, in <module>
>     from . import certs
> ImportError: cannot import name certs
> [DEBUG   ] Module DSC: Only available on Windows systems
> [DEBUG   ] Module PSGet: Only available on Windows systems
> [DEBUG   ] Could not LazyLoad mysql.user_create: 'mysql' __virtual__ returned False: The mysql execution module cannot be loaded: neither MySQLdb nor PyMySQL is available.
> [DEBUG   ] LazyLoaded acme.cert
> [DEBUG   ] Could not LazyLoad apache.a2enconf: 'apache.a2enconf' is not available.
> [DEBUG   ] Could not LazyLoad apache.a2enmod: 'apache.a2enmod' is not available.
> [DEBUG   ] Could not LazyLoad apache.a2ensite: 'apache.a2ensite' is not available.
> [DEBUG   ] Could not LazyLoad pkg.get_selections: 'pkg.get_selections' is not available.
> [DEBUG   ] Could not LazyLoad augeas.execute: 'augeas.execute' is not available.
> [DEBUG   ] LazyLoaded bigip.list_transaction
> [DEBUG   ] LazyLoaded boto3_elasticache.cache_cluster_exists
> [DEBUG   ] LazyLoaded boto3_route53.find_hosted_zone
> [DEBUG   ] LazyLoaded boto_apigateway.describe_apis
> [DEBUG   ] LazyLoaded boto_asg.exists
> [DEBUG   ] LazyLoaded boto_cfn.exists
> [DEBUG   ] LazyLoaded boto_cloudtrail.exists
> [DEBUG   ] LazyLoaded boto_cloudwatch.get_alarm
> [DEBUG   ] LazyLoaded boto_cloudwatch_event.exists
> [DEBUG   ] LazyLoaded boto_cognitoidentity.describe_identity_pools
> [DEBUG   ] LazyLoaded boto_datapipeline.create_pipeline
> [DEBUG   ] LazyLoaded boto_dynamodb.exists
> [DEBUG   ] LazyLoaded boto_ec2.get_key
> [DEBUG   ] LazyLoaded boto_elasticache.exists
> [DEBUG   ] LazyLoaded boto_elasticsearch_domain.exists
> [DEBUG   ] LazyLoaded boto_elb.exists
> [DEBUG   ] LazyLoaded boto_elbv2.target_group_exists
> [DEBUG   ] LazyLoaded boto_iam.get_user
> [DEBUG   ] LazyLoaded boto_iam.role_exists
> [DEBUG   ] LazyLoaded boto_iot.policy_exists
> [DEBUG   ] LazyLoaded boto_kinesis.exists
> [DEBUG   ] LazyLoaded boto_kms.describe_key
> [DEBUG   ] LazyLoaded boto_lambda.function_exists
> [DEBUG   ] LazyLoaded boto_asg.exists
> [DEBUG   ] LazyLoaded boto_rds.exists
> [DEBUG   ] LazyLoaded boto_route53.get_record
> [DEBUG   ] LazyLoaded boto_s3_bucket.exists
> [DEBUG   ] LazyLoaded boto_secgroup.exists
> [DEBUG   ] LazyLoaded boto_sns.exists
> [DEBUG   ] LazyLoaded boto_sqs.exists
> [DEBUG   ] LazyLoaded boto_vpc.exists
> [DEBUG   ] LazyLoaded bower.list
> [DEBUG   ] LazyLoaded chef.client
> [DEBUG   ] LazyLoaded chocolatey.install
> [DEBUG   ] LazyLoaded cisconso.set_data_value
> [DEBUG   ] LazyLoaded cyg.list
> [DEBUG   ] LazyLoaded ddns.update
> [DEBUG   ] LazyLoaded chassis.cmd
> [DEBUG   ] Could not LazyLoad docker.version: 'docker.version' is not available.
> [DEBUG   ] Could not LazyLoad docker.version: 'docker.version' is not available.
> [DEBUG   ] Could not LazyLoad docker.version: 'docker.version' is not available.
> [DEBUG   ] Could not LazyLoad docker.version: 'docker.version' is not available.
> [DEBUG   ] Could not LazyLoad docker.version: 'docker.version' is not available.
> [DEBUG   ] LazyLoaded eselect.exec_action
> [DEBUG   ] LazyLoaded esxi.cmd
> [DEBUG   ] LazyLoaded git.version
> [DEBUG   ] LazyLoaded github.list_users
> [DEBUG   ] LazyLoaded glusterfs.list_volumes
> [DEBUG   ] LazyLoaded elasticsearch.exists
> [DEBUG   ] LazyLoaded grafana4.get_dashboard
> [DEBUG   ] LazyLoaded grafana4.get_datasource
> [DEBUG   ] Failed to import states grafana4_org:
> Traceback (most recent call last):
>   File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1411, in _load_module
>     mod = imp.load_module(mod_namespace, fn_, fpath, desc)
>   File "/usr/lib/python2.7/site-packages/salt/states/grafana4_org.py", line 48, in <module>
>     from requests.exceptions import HTTPError
>   File "/usr/lib/python2.7/site-packages/requests/__init__.py", line 58, in <module>
>     from . import utils
>   File "/usr/lib/python2.7/site-packages/requests/utils.py", line 25, in <module>
>     from . import certs
> ImportError: cannot import name certs
> [DEBUG   ] LazyLoaded grafana4.get_user
> [DEBUG   ] Failed to import states grafana_dashboard:
> Traceback (most recent call last):
>   File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1411, in _load_module
>     mod = imp.load_module(mod_namespace, fn_, fpath, desc)
>   File "/usr/lib/python2.7/site-packages/salt/states/grafana_dashboard.py", line 45, in <module>
>     import requests
>   File "/usr/lib/python2.7/site-packages/requests/__init__.py", line 58, in <module>
>     from . import utils
>   File "/usr/lib/python2.7/site-packages/requests/utils.py", line 25, in <module>
>     from . import certs
> ImportError: cannot import name certs
> [DEBUG   ] Failed to import states grafana_datasource:
> Traceback (most recent call last):
>   File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1411, in _load_module
>     mod = imp.load_module(mod_namespace, fn_, fpath, desc)
>   File "/usr/lib/python2.7/site-packages/salt/states/grafana_datasource.py", line 29, in <module>
>     import requests
>   File "/usr/lib/python2.7/site-packages/requests/__init__.py", line 58, in <module>
>     from . import utils
>   File "/usr/lib/python2.7/site-packages/requests/utils.py", line 25, in <module>
>     from . import certs
> ImportError: cannot import name certs
> [DEBUG   ] LazyLoaded icinga2.generate_ticket
> [DEBUG   ] LazyLoaded ifttt.trigger_event
> [DEBUG   ] Could not LazyLoad influxdb08.db_exists: 'influxdb08.db_exists' is not available.
> [DEBUG   ] Could not LazyLoad influxdb08.db_exists: 'influxdb08.db_exists' is not available.
> [DEBUG   ] Could not LazyLoad influxdb.db_exists: 'influxdb.db_exists' is not available.
> [DEBUG   ] Could not LazyLoad influxdb.db_exists: 'influxdb.db_exists' is not available.
> [DEBUG   ] Could not LazyLoad influxdb.db_exists: 'influxdb.db_exists' is not available.
> [DEBUG   ] Could not LazyLoad influxdb.db_exists: 'influxdb.db_exists' is not available.
> [DEBUG   ] LazyLoaded infoblox.get_record
> [DEBUG   ] LazyLoaded ipset.version
> [DEBUG   ] LazyLoaded kapacitor.version
> [DEBUG   ] LazyLoaded keystone.auth
> [DEBUG   ] LazyLoaded kubernetes.ping
> [DEBUG   ] LazyLoaded layman.add
> [DEBUG   ] LazyLoaded lvs.get_rules
> [DEBUG   ] LazyLoaded lvs.get_rules
> [DEBUG   ] LazyLoaded makeconf.get_var
> [DEBUG   ] LazyLoaded memcached.status
> [DEBUG   ] LazyLoaded mongodb.user_exists
> [DEBUG   ] LazyLoaded monit.summary
> [DEBUG   ] LazyLoaded mysql.db_exists
> [DEBUG   ] LazyLoaded mysql.grant_exists
> [DEBUG   ] LazyLoaded mysql.query
> [DEBUG   ] LazyLoaded nftables.version
> [DEBUG   ] LazyLoaded npm.list
> [DEBUG   ] LazyLoaded nxos.cmd
> [DEBUG   ] LazyLoaded openvswitch.bridge_create
> [DEBUG   ] LazyLoaded openvswitch.port_add
> [DEBUG   ] LazyLoaded pagerduty_util.get_resource
> [DEBUG   ] LazyLoaded pagerduty_util.get_resource
> [DEBUG   ] LazyLoaded pagerduty_util.get_resource
> [DEBUG   ] LazyLoaded pagerduty_util.get_resource
> [DEBUG   ] LazyLoaded pecl.list
> [DEBUG   ] LazyLoaded portage_config.get_missing_flags
> [DEBUG   ] LazyLoaded postgres.cluster_exists
> [DEBUG   ] LazyLoaded postgres.user_exists
> [DEBUG   ] LazyLoaded postgres.create_extension
> [DEBUG   ] LazyLoaded postgres.group_create
> [DEBUG   ] LazyLoaded postgres.datadir_init
> [DEBUG   ] LazyLoaded postgres.language_create
> [DEBUG   ] LazyLoaded postgres.privileges_grant
> [DEBUG   ] LazyLoaded postgres.schema_exists
> [DEBUG   ] LazyLoaded postgres.tablespace_exists
> [DEBUG   ] LazyLoaded postgres.user_exists
> [DEBUG   ] Could not LazyLoad rbac.profile_list: 'rbac.profile_list' is not available.
> [DEBUG   ] LazyLoaded rdp.enable
> [DEBUG   ] Could not LazyLoad redis.set_key: 'redis.set_key' is not available.
> [DEBUG   ] LazyLoaded reg.read_value
> [DEBUG   ] Could not LazyLoad vmadm.create: 'vmadm.create' is not available.
> [DEBUG   ] LazyLoaded snapper.diff
> [DEBUG   ] LazyLoaded splunk.list_users
> [DEBUG   ] LazyLoaded splunk_search.get
> [DEBUG   ] LazyLoaded stormpath.create_account
> [DEBUG   ] LazyLoaded telemetry.get_alert_config
> [DEBUG   ] LazyLoaded tls.cert_info
> [DEBUG   ] LazyLoaded tomcat.status
> [DEBUG   ] LazyLoaded trafficserver.set_config
> [DEBUG   ] LazyLoaded uptime.checks_list
> [DEBUG   ] LazyLoaded victorops.create_event
> [DEBUG   ] LazyLoaded virt.node_info
> [DEBUG   ] LazyLoaded win_dacl.add_ace
> [DEBUG   ] LazyLoaded win_dns_client.add_dns
> [DEBUG   ] Could not LazyLoad firewall.get_config: 'firewall.get_config' is not available.
> [DEBUG   ] LazyLoaded win_iis.create_site
> [DEBUG   ] Could not LazyLoad lgpo.set: 'lgpo.set' is not available.
> [DEBUG   ] LazyLoaded win_path.rehash
> [DEBUG   ] LazyLoaded win_pki.get_stores
> [DEBUG   ] LazyLoaded win_servermanager.install
> [DEBUG   ] LazyLoaded win_smtp_server.get_server_setting
> [DEBUG   ] LazyLoaded win_snmp.get_agent_settings
> [DEBUG   ] LazyLoaded x509.get_pem_entry
> [DEBUG   ] LazyLoaded xmpp.send_msg
> [DEBUG   ] LazyLoaded zabbix.host_create
> [DEBUG   ] LazyLoaded zabbix.hostgroup_create
> [DEBUG   ] LazyLoaded zabbix.mediatype_create
> [DEBUG   ] LazyLoaded zabbix.user_create
> [DEBUG   ] LazyLoaded zabbix.usergroup_create
> [DEBUG   ] LazyLoaded zenoss.add_device
> [WARNING ] salt.loaded.int.states.zenoss.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'zenoss', please fix this.
> [DEBUG   ] LazyLoaded zfs.create
> [DEBUG   ] LazyLoaded zk_concurrency.lock
> [DEBUG   ] LazyLoaded zonecfg.create
> [DEBUG   ] LazyLoaded zpool.create
> [DEBUG   ] Could not LazyLoad mysql_user.present: 'mysql_user' __virtual__ returned False
> [INFO    ] Running state [frank] at time 10:20:18.387925
> [DEBUG   ] LazyLoaded mysql_user.present
> [ERROR   ] State 'mysql_user.present' was not found in SLS 'test'
> Reason: 'mysql_user' __virtual__ returned False
> 
> [DEBUG   ] File /var/cache/salt/minion/accumulator/140128415733008 does not exist, no need to cleanup.
> [DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'aspod-johan-main01', 'tcp://salt:4506', 'aes')
> [DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', 'aspod-johan-main01', 'tcp://salt:4506')
> [DEBUG   ] LazyLoaded highstate.output
> local:
>           ID: frank
>     Function: mysql_user.present
>       Result: False
>      Comment: State 'mysql_user.present' was not found in SLS 'test'
>               Reason: 'mysql_user' __virtual__ returned False
>      Changes:
> 
> Summary for local
> ------------
> Succeeded: 0
> Failed:    1
> ------------
> Total states run:     1
> Total run time:   0.000 ms

### Steps to Reproduce Issue

1. Latest salt-master on Ubuntu 16.04
2. Latest salt-minion on Red Hat 6
3. Install MySQL (mysql  Ver 14.14 Distrib 5.1.73, for redhat-linux-gnu (x86_64) using readline 5.1)
4. Run the state above from the minion: salt-call state.sls test -l debug

### Versions Report
Minion
```
Salt Version:
           Salt: 2017.7.1

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.8.1
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.13 (default, Jul 12 2017, 17:32:34)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.5.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5

System Versions:
           dist: redhat 6.9 Santiago
         locale: UTF-8
        machine: x86_64
        release: 2.6.32-696.6.3.el6.x86_64
         system: Linux
        version: Red Hat Enterprise Linux Server 6.9 Santiago
```

Master
```
Salt Version:
           Salt: 2017.7.1

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.6.0
      docker-py: 2.2.1
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.3
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: 1.3.7
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.12 (default, Nov 19 2016, 06:48:10)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: Ubuntu 16.04 xenial
         locale: UTF-8
        machine: x86_64
        release: 4.4.0-1031-aws
         system: Linux
        version: Ubuntu 16.04 xenial
```

**Comment by Ch3LL** [comment]
Do you have the MySQLdb python27 module installed? in 2017.7.0 we dropped python 2.6 support so salt will only run on python2.7 in redhat6.

**Comment by jryberg** [comment]
Hi @Ch3LL, thank you for your reply. 

Python 2.7 are not supported by Red Hat 6

Does this mean that SaltStack are no longer supporting Red Hat/CentOS 6?

I know it's technically possible to install Python 2.7 but It's not supported by Red Hat and as we are using Red Hat 6 in production environments we are not able to do anything that are unsupported by Red Hat. 

Best regards Johan Ryberg

**Comment by jryberg** [comment]
This is a how to install "python27-mysql" on Red Hat 6

- yum -y install gcc
- yum -y install rpm-build gcc-c++ # not entire sure this is needed, have not tested without them
- yum -y install mysql-devel python27-devel
- curl https://bootstrap.pypa.io/get-pip.py | python2.7 -
- pip2.7 install MySQL-python

I suppose I will have to create a rpm package of product for our environments but I'm afraid the move from python 2.6 to 2.7 will cause many issues for us who has not migrated away from Red Hat 6 / CentOS 6

**Comment by jryberg** [comment]
Guide how to create rpm for redistribution
```
  yum install ruby-devel gcc curl libyaml-devel
  command curl -sSL https://rvm.io/mpapis.asc | sudo gpg2 --import -
  curl -L get.rvm.io | bash -s stable
  source /etc/profile.d/rvm.sh
  rvm requirements
  rvm install 1.9.3
  rvm use 1.9.3 --default
  rvm rubygems current
  gem install fpm
  fpm --debug -s python -t rpm --python-bin python2.7 --python-install-lib /usr/lib/python2.7/site-packages --rpm-dist 'el6' mysql-python
```

**Comment by Ch3LL** [comment]
@jryberg redhat6 is still supported by saltstack so we have provided python27 packages in our repository for those that need to install python27 alongside centos6 to workaround not supporting python2.6 any longer. Docs are [here](https://docs.saltstack.com/en/latest/topics/releases/2017.7.0.html#python-2-6-deprecation) in the release notes. In fact it looks like it already installed for you as your version report shows python version 2.7. Thanks for posting the instructions to build the rpm, might help out someone else who is running into the issue. do you have any other concerns?

**Comment by jryberg** [comment]
Hi @Ch3LL 

I do not wish to be rude of your work to make SaltStack better, I understand that Python 2.6 are very old and are causing issues for maintainability of the software.

You are correct that you are bundling Python 2.7 and you are correct that salt **core** functionality works.

I do have one very strong concern that almost all modules, states and other components that requires external python libraries will fail after 2017 version since there are no official builds for most of them for Red Hat 6 and Python 2.7 since they all depends on Python 2.6

As you can see in this doc:
https://docs.saltstack.com/en/latest/ref/states/all/salt.states.mysql_user.html

> depends: MySQLdb Python module

This state are officially supported by SaltStack for all supported platforms but it will no longer work

I just assume all states or other components that have a dependency to any Python library stopped working. 

I really appreciate all the work this community does but I can't stop feel that Red Hat 6 are no longer "fully" supported, it's now degraded since standard components that used to work are no longer working and libraries that was available are no longer available since Red Hat or EPEL are bundling Python libraries for 2.7, only for 2.6.

I don't know what to do with this issue, there are no turning back but I would like to have someone confirming my concerns and findings. 

Best regards Johan Ryberg

**Comment by davidjb** [comment]
Fwiw, can confirm the same problem with v2016.11.6 of Salt; eg not just v2017+.

**Comment by moench-tegeder** [comment]
I have to agree that I'm a little surprised at the loss of saltstack's "optional" functionality in Redhat/CentOS 6 based systems. Perhaps the remark in the release notes could be put in some stronger words? It's not only the mysql (and some other databases) states (just checking :depends: tags), but also e.g. x509 (requires M2Crypto and about the whole schedule subsystem (without dateutil and/or croniter there's not that much left...).
Rolling my own python packages looks like much more work than I expected on this old machine (already marked for migration).

**Comment by Ch3LL** [comment]
Thank you for laying out your concerns guys. 
One way to handle these additional dependencies is using `pip2.7 install < package > ` as well or the `pip.installed` state. Is that not working as well or maybe lay out why this approach is not vaiable? I believe you will first need to install the python2.7-pip package first. (i'm not entirely positive the exact name of the package)

I can definitely update the release notes with finer detail around this additional requirement to install these additional dependencies and mention the pip install method as well but want to get some feedback from you as well.

**Comment by moench-tegeder** [comment]
Ah, there's a python27-pip, completely missed that package. Thanks a lot.
Perhaps that's the thing to mention in the release notes, somewhat along the lines of "Additional python modules required for some functionality can be installed via pip2.7 (package python-pip27)".  I was already collecting stuff from setuptools upwards...

This should be good as a workaround - I'm not planning on keeping that machine around for much longer :)

**Comment by jryberg** [comment]
I'm working in almost 100% sealed of environments and every single package needs to be imported to a special repository. 

pip does not work since pip need Internet access to public repositories. We will have to convert pip to rpm for every single package we need and also maintain it since we will not get security updates this way. 

It's no concerns, it's a real fact that something very important was removed without proper investigation how it in real life will impact end users. 

The best thing would be to drop support entirely for new versions and keep backward comparability for older until RHEL6 are EOL

**Comment by damon-atkins** [comment]
I would suggest Salt team need to create more RPM's for python 2.7 on old OS for to support mysql and other items that have python interfaces and are part of salt modules.
e.g.  Install RHEL 6, and compile up the extra python 2.7 modules which currently come with RHEL python2.6 that are need by all salt modules. It would give a better customer experience.

**Comment by lorengordon** [comment]
@jryberg No need to convert python packages to rpm. You can use pip to create a local repo, then distribute the local repo and install the packages and their dependencies from that.

https://stackoverflow.com/a/22093005

You can also re-host the repo on an internal web server as a custom index server, but that's a little trickier.

**Comment by robthralls** [comment]
I would just like to voice my frustration with SaltStack as well.

I found this thread because the "virt" execution module is now broken on EL6, since the libvirt-python package is for Python 2.6 and SaltStack does not provide a 2.7 package.

I suspect I'm probably in a similar situation as @jryberg. Due to contractual/legal requirements placed on my environments such as package signing and key review, decentralized repositories like pip are NOT an acceptable alternative. I really would need to maintain my own RPM packages for anything SaltStack now depends on but does not provide.

I do also appreciate all the work that's gone into this project, but I'm honestly getting a bit fed up with the terrible end user experience. It's the same story every time we're backed into a corner and forced to update SaltStack; something major is (understandably) changed, but seemingly no regression testing is done, nor does there seem to be any consideration given to the wider support impact.

Similar to this incident were the recent crypto changes. I run FIPS environments. SaltStack has _repeatedly_ been completely broken for us. I have personally opened up a couple of tickets and from what I remember, there has always been a relatively quick workaround, but I still fully expect to run into FIPS issues every time I update.

And that's basic functionality. The "optional" modules are a whole other story. The lesson I've had repeatedly beaten into me is that while SaltStack has hundreds of modules... the vast majority of them are unusable garbage. When they aren't constantly being broken, the functionality they would provide is useless. You're generally much better served by managing the files directly or issuing the system commands yourself. Because that's usually what the modules are lazily doing anyway, often with incorrect or incomplete assumptions. If you stray from a few "core" functions selected from a few "core" modules, you WILL be burned. Every time.

End rant.

**Comment by moench-tegeder** [comment]
Wondering - why try to run the latest&greatest SaltStack release on that old platform anyways? Given. this issue has some surprise factor - but shouldn't there be any local pre-rollout-testing? (I'm guilty on that one too - "that one machine long past best-before"...). Sorry in case this comes across somewhat snarky, but I've seen environments with a huge "change process" with a lot of sign-offs even for simple things - but with all the process, the single thing which didn't happen was testing the change. That triggers rants, too :)

**Comment by robthralls** [comment]
> shouldn't there be any local pre-rollout-testing?

Oh sure. You still have to spend time dealing with the issues you find, though.

> why try to run the latest&greatest SaltStack release on that old platform anyways?

Because it's easier. No, seriously. SaltStack doesn't have an LTS release and CVE support ends after only 18 months. Stuff like CVE-2017-12791 last month comes up on the scans we are required to report on. I am forced to either update or else patch it myself, provide evidences that the vulnerability has actually been remediated, and provide a write-up justifying both the unsupported release as operationally required and the scan result as a false positive. A hard sell even _with_ a compelling reason. I suspect I would sooner be told to remove SaltStack entirely if it can't be updated.

Anyways.... Sorry to unload on you all.

**Comment by Ch3LL** [comment]
I just added a note to the nitrogen release notes to mention the pip workaround to the CentOS6 dependencies for modules. As stated before this is our current workaround and we don't intend on packaging other modules then what is available currently. Python 2.6 is deprecated and we intend to keep it deprecated on our platform to try to align with the upstream, even though RedHat has chosen not to do the same with CentOS6.

As a side note around the fips issue @robthralls we actually added fips to our auto package testing since we noticed your issues this most recent release, so hopefully that helps with fips compatibility in the future.
